### DISCUSSION TITLE
支持gemini吗？

### DISCUSSION THREAD

**Discussion by CN-Steve-Lee** [KenyonY/openai-forward#105] [Q&A] [answered]
支持gemini API的转发吗

**Comment by KenyonY** [comment]
刚刚更新了README: [代理其它云端模型](https://github.com/KenyonY/openai-forward?tab=readme-ov-file#%E4%BB%A3%E7%90%86%E5%85%B6%E5%AE%83%E4%BA%91%E7%AB%AF%E6%A8%A1%E5%9E%8B)  

原则上使用`EXTRA_BASE_URL`可以转发任何服务至`EXTRA_ROUTE_PREFIX`路由前缀上（需要注意不要和默认的`OPENAI_ROUTE_PREFIX`冲突），但是不支持自定义秘钥、实时日志和缓存AI响应等功能。 
可以升级到最新版本试试看

**Reply by CN-Steve-Lee** [reply]
感谢回复，因为没有示例，我还是有点不清楚EXTRA_BASE_URL和EXTRA_ROUTE_PREFIX的作用。
我有一个80端口运行了openai的转发(aifd run --port=80),我如果想再转发gemini的话，是要再在另一个目录开个比如8080端口的gemini的服务(aifd run --port=8080 ，.env里写的#OPENAI_ROUTE_PREFIX='/v1beta/models/gemini-pro:generateContent，测试过可以成功转发gemini api)，然后再在openai的那个env文件中的EXTRA_BASE_URL写='localhost:8080'才可以实现通过访问80端口就既可以转发openai也可以转发gemini的吗？
还是说我对EXTRA_BASE_URL理解有误？
(我的需求是转发openai和gemini的api，如果能通过一个80端口来转发2个就最好)

同时还有个小小的问题，关于能不能通过运行python脚本来代替aifd run，这样的话就可以直接用云函数部署了，可能开发成本更低？(我没搞定转写aifd run才开了一台服务器干这个事hhh)

最后真的很感谢你的回复和你的项目，很有价值！

**Reply by KenyonY** [reply, marked as answer]
确实这个项目的文档写得比较糟糕...   
1. 可以同时转发多个服务，只需要将这些服务放置在对应的路由上就可以，下面是示例：
```sh
OPENAI_BASE_URL='http://xxx1, http://xxx2'
OPENAI_ROUTE_PREFIX='/openai1, /openai2'
```
使用`aifd run`启动时，就可以在`/openai1`访问到http://xxx1 的内容，在`/openai2`访问到 http://xxx2 的内容。  
`OPENAI_BASE_URL`与 `EXTRA_BASE_URL`的区别主要是在对转发内容的解析上，`EXTRA_BASE_URL`不对转发内容进行解析和定制，可以转发任何内容，所以建议对非openai格式的服务`EXTRA_BASE_URL`来转发。

2. 使用 [deploy to render](https://github.com/KenyonY/openai-forward/blob/main/deploy.md#render-%E4%B8%80%E9%94%AE%E9%83%A8%E7%BD%B2) 进行转发呀，只要在render的配置中填写相关环境变量就可以了。render上除了免费用户无法保存文件这个限制，其它功能都可以实现。    
  如果希望用python启动也是可以的，`aifd run` 等同于 `python -m openai_forward.__main__ run`

**Reply by CN-Steve-Lee** [reply]
感谢慷慨解惑！对我们lab很有价值！
